python/ray/tune/schedulers/resource_changing_scheduler.py

```diff
@@ -0,0 +1,222 @@
+from ray.tune.utils.placement_groups import PlacementGroupFactory
+
+logger = logging.getLogger(__name__)
+
+
+class ResourceChangingScheduler(TrialScheduler):
+    """A utility scheduler to dynamically change resources of live trials.
+
+    Experimental. API may change in future releases.
+
+    The ResourceChangingScheduler works by wrapping around any other
+    scheduler and adjusting the resource requirements of live trials
+    in response to the decisions of the wrapped scheduler
+    through a user-specified `resources_allocation_function`. An example
+    of such a function can be found in
+    `tune/examples/xgboost_dynamic_resources_example.py`.
+
+    Only supports the Trainable (class) API. If resources of a trial are
+    updated with new values, the `update_resources` method in the Trainable
+    will be called. This method needs to be overwritten by the user in
+    order to let the trained model take advantage of newly allocated resources.
+
+    Cannot be used if `reuse_actors` is True in `tune.run`. A ValueError
+    will be raised in that case.
+
+    Args:
+        base_scheduler (TrialScheduler): The scheduler to provide decisions
+            about trials. If None, a default FIFOScheduler will be used.
+        resources_allocation_function (Callable): The function used to change
+            live trial resource requiements during tuning. This function
+            will be called on each trial as it finishes one step of training.
+            The function must take four arguments: `TrialRunner`, current
+            `Trial`, current result `dict` and the base trial resource
+            `PlacementGroupFactory` or `Resource` (depending on whether
+            placement groups are used). The function must return a
+            `PlacementGroupFactory`, `Resources`,`dict` or None (signifying
+            no need for an update). If `resources_allocation_function` is
+            None, no resource requirements will be changed at any time.
+
+    Warning:
+        If the `resources_allocation_function` sets trial resource requirements
+        to values bigger than possible, the trial will not run. Ensure
+        that your function accounts for that possibility by setting upper
+        limits. Consult the example file to see how that may be done.
+
+    Example:
+        >>> base_scheduler = ASHAScheduler(max_t=16)
+        >>> def resource_allocation_function(
+        >>>     trial_runner: "trial_runner.TrialRunner",
+        >>>     trial: Trial,
+        >>>     result: dict,
+        >>>     base_trial_resource: Union[PlacementGroupFactory|Resource]
+        >>> ) -> Union[None, PlacementGroupFactory, Resource]:
+        >>>     # logic here
+        >>>     # usage of PlacementGroupFactory is strongly preferred
+        >>>     return PlacementGroupFactory(...)
+        >>> scheduler = ResourceChangingScheduler(base_scheduler,
+        >>>                                     resource_allocation_function)
```

Comment:
`.. code-block:: python`?

Reply:
It seems to render fine - I've seen it done like that in some other docstrings. I think it also plays better with IDEs - using `code-block` didn't render anything in VSCode.
<img width="730" alt="Screenshot 2021-07-09 at 13 33 20" src="https://user-images.githubusercontent.com/10364161/125071972-4587d200-e0ba-11eb-9673-0f9802fe4db3.png">

Reply:
That being said it should probably be a code-block for consistency